### PR TITLE
CE-95: event date

### PR DESCRIPTION
**Pull request by ialarmedalien**
New event date class, all fields mandatory.

`date`: will be parsed by the credit engine and then formatted as YYYY, YYYY-MM, or YYYY-MM-DD, depending on the level of specificity of the input
`dateFormat`: string literal representing the date; will be `YYYY`, `YYYY-MM`, or `YYYY-MM-YY`
`event`: string describing what the event was -- e.g. dataset created, dataset published, dataset embargo lifted, etc.

### CODE REVIEW

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [inline src/us/kbase/workspace/database/provenance/EventDate.java:27]: indent issue?

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [inline src/us/kbase/workspace/database/provenance/EventDate.java:19]: Could we have some docs for this regex please? It's pretty gnarly to read

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [inline src/us/kbase/workspace/database/provenance/EventDate.java:33]: I'm forgetting why a `Date` or better, `Instant` instance wouldn't work here

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [inline src/us/kbase/workspace/database/provenance/EventDate.java:47]: Is there any controlled vocab here?

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [inline src/us/kbase/workspace/database/provenance/EventDate.java:85]: It's possible for the date to not match the date format - that seems like it should be checked here, unless switching to a class instance for the date in which case the date format essentially acts like a sigfigs spec

**Review by MrCreosote — CHANGES_REQUESTED** [review]

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (ialarmedalien) [reply src/us/kbase/workspace/database/provenance/EventDate.java:33]: In some cases, we only have the year; in some just the date; and in some, the full date + time.

Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/EventDate.java:33]: But it could still be represented as a Date/Instant

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (ialarmedalien) [reply src/us/kbase/workspace/database/provenance/EventDate.java:85]: Crap, I meant to add a check for that but then forgot about it. :facepalm:

I was mulling over whether to use one of the date modules to parse the input, which would give better validation than just using a regex, and you could ensure that the `date` and `dateFormat` fields were consistent. Presumably I would have to use `Date`, rather than `Instant`, since we don't have millisecond (or in some cases, month!) precision for the dates to be captured.

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (ialarmedalien) [reply src/us/kbase/workspace/database/provenance/EventDate.java:47]: Potentially, although I need to discuss this further with Elisha _et al._  at an upcoming CE meeting. For now I will leave as-is.

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/EventDate.java:85]: > better validation than just using a regex

that's where my head's at

>  Date, rather than Instant, since we don't have millisecond (or in some cases, month!) precision 

`Instant` should be fine with that. It can do anything `Date` can do (except be mutated). You shouldn't use `Date` if at all possible

**Comment by MrCreosote** [comment]
That force push just added a bunch of stuff from a prior PR to this PR

**Comment by ialarmedalien** [comment]
yep, it's out of date with respect to the base branch. I will fix it once the nitpick PR is merged to develop.

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (ialarmedalien) [reply src/us/kbase/workspace/database/provenance/EventDate.java:85]: According to https://docs.oracle.com/javase/tutorial/datetime/iso/overview.html, `Instant` deals with (nano)seconds, and the constructor uses epoch time, so it's not going to be suitable for dates like `2022`, where only the year is available. The `date` field would need to be a `Year`, `LocalDate` (YMD), or possibly a `YearMonth`, to deal with the variations on the representation of the date that we have.

I am tempted to go back to storing the date as a string but validate it by attempting to convert it into either a `LocalDate`, a `YearMonth`, or a `Year`, and then normalise it by outputting it using a known `DateTimeFormatter` (e.g. `YYYY-MM`, `YYYY`, or `DateTimeFormatter.ISO_LOCAL_DATE`). 

Your thoughts?

Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/EventDate.java:85]: > constructor uses epoch time, so it's not going to be suitable for dates like 2022, where only the year is available. 

I'm not understanding why that is - you can certainly represent 2022 in epoch time. `Date` also uses epoch time under the hood.

```
final LocalDate ld = LocalDate.ofYearDay(2022, 1);
final Instant inst = ZonedDateTime.of(ld, LocalTime.MIN, ZoneId.of("Z")).toInstant();
System.out.println(inst);
```
prints
```
2022-01-01T00:00:00Z
```

File: src/us/kbase/workspace/database/provenance/Event.java
Comment (MrCreosote) [inline src/us/kbase/workspace/database/provenance/Event.java:42]: This doc seems off

**Review by MrCreosote — CHANGES_REQUESTED** [review]
I'm assuming here that nitpickery will be removed and this branch will be retargeted to develop

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/EventDate.java]: :+1:

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/EventDate.java]: I think this is moot at this point

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/EventDate.java:25]: IOW DRTW

File: src/us/kbase/workspace/database/provenance/Event.java
Comment (ialarmedalien) [reply src/us/kbase/workspace/database/provenance/Event.java:24]: Oooops, c&p error.

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (ialarmedalien) [reply src/us/kbase/workspace/database/provenance/EventDate.java:25]: That's a worrying thought. I would have hoped that humankind were extinct by then.

File: src/us/kbase/workspace/database/provenance/Event.java
Comment (ialarmedalien) [reply src/us/kbase/workspace/database/provenance/Event.java:42]: this was a really shoddy PR! Thank goodness for code review.

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/EventDate.java:79]: I don't think I understand

Comment (ialarmedalien) [reply src/us/kbase/workspace/database/provenance/EventDate.java:79]: I think I have misunderstood your comment -- can you write a code sample to illustrate?

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/EventDate.java:25]: Always the optimist

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/EventDate.java:79]:
```java
public static EventDate build(final String date, final String event) {
    doStuff();
    return EventDate(date, eventEnum);
}
```
and an equivalent on for an enum event (and the former could call the latter, in fact)

Comment (ialarmedalien) [reply src/us/kbase/workspace/database/provenance/EventDate.java:79]: but no `whatever.getBuilder(...)`? That's what I meant about inconsistent -- having some classes where you have to call `blah.getBuilder(...)` and others where you can just call `blah.build(...)`

Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/EventDate.java:79]: > but no whatever.getBuilder(...)?

Right

> That's what I meant about inconsistent -- having some classes where you have to call blah.getBuilder(...).build() and others where you can just call blah.build(...)

Pretty standard in java land TBH. Use a builder when necessary (typically for optional arguments) but if it's not just use a static create method. Don't force 2 method calls when one is effectively redundant

Comment (ialarmedalien) [reply src/us/kbase/workspace/database/provenance/EventDate.java:79]: Fair enough.

File: src/us/kbase/workspace/database/provenance/Common.java
Comment (ialarmedalien) [inline src/us/kbase/workspace/database/provenance/Common.java:92]: no longer needed as no longer using a regex

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/EventDate.java]: :+1:

File: src/us/kbase/workspace/test/database/provenance/EventDateTest.java
Comment (MrCreosote) [reply src/us/kbase/workspace/test/database/provenance/EventDateTest.java:49]: :+1:

File: src/us/kbase/workspace/test/database/provenance/EventDateTest.java
Comment (MrCreosote) [reply src/us/kbase/workspace/test/database/provenance/EventDateTest.java]: :+1:

File: src/us/kbase/workspace/database/provenance/Event.java
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/Event.java:42]: :+1:

File: src/us/kbase/workspace/database/provenance/Event.java
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/Event.java]: :+1:

File: src/us/kbase/workspace/database/provenance/Event.java
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/Event.java]: :+1:

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/EventDate.java]: :+1:

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/EventDate.java]: moot now

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/EventDate.java]: :+1:

File: src/us/kbase/workspace/test/database/provenance/EventTest.java
Comment (MrCreosote) [reply src/us/kbase/workspace/test/database/provenance/EventTest.java:52]: :+1:

File: src/us/kbase/workspace/database/provenance/Event.java
Comment (MrCreosote) [inline src/us/kbase/workspace/database/provenance/Event.java:50]:
```suggestion
	 * @param input a string representing an event.
```

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [inline src/us/kbase/workspace/database/provenance/EventDate.java:70]:
```suggestion
	 * @return the new {@link EventDate}.
```
Seems trivial but the main benefit is that auto refactoring can change the name here if the class name is refactored

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [inline src/us/kbase/workspace/database/provenance/EventDate.java:83]:
```suggestion
	 * @return the new {@link EventDate}.
```

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [inline src/us/kbase/workspace/database/provenance/EventDate.java:69]: Nit^10: Do you want to mention the string has to match an enum value?

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (MrCreosote) [inline src/us/kbase/workspace/database/provenance/EventDate.java:114]: This is probably ok, but the error message might be confusing for dates that are formatted correctly but invalid, like 2022-04-31. It seems improbable that that'll get passed, so as long as it's not accepted this is probably good enough, but I leave it up to you.

**Review by MrCreosote — APPROVED** [review]
Minor stuff, up to you what you want to do about it.

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (ialarmedalien) [reply src/us/kbase/workspace/database/provenance/EventDate.java:114]: I'll edit the message to mention that the date must be valid.

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (ialarmedalien) [reply src/us/kbase/workspace/database/provenance/EventDate.java:69]: Will do.

File: src/us/kbase/workspace/database/provenance/EventDate.java
Comment (ialarmedalien) [reply src/us/kbase/workspace/database/provenance/EventDate.java:115]: If I am feeling energetic at some future time point I'll write something to parse out the DateTimeParseException error. I am not feeling energetic ATM.

Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/EventDate.java:115]: I seriously doubt it's worth the effort ever

**Pull request merged**
